Security: Fix Broken Access Control Allowing Non-Superusers to Edit Frontmatter

nakkouchtarek opened the pull request:
Due to a weak check of authorization in the `taskSave` method of `AdminController.php`, an editor-level user was able to modify the frontmatter of a form. This allowed them to enable Twig and modify how the form processes form data.

This fix introduces a permission check to ensure that only superusers are allowed to edit the frontmatter field.

An advisory has already been opened regarding this issue :

[https://github.com/getgrav/grav/security/advisories/GHSA-v8x2-fjv7-8hjh](https://github.com/getgrav/grav/security/advisories/GHSA-v8x2-fjv7-8hjh)

nakkouchtarek committed:
Fixed non-superuser from editing frontmatter

rhukster reviewed (changes requested):
## Review: Security Fix for Frontmatter Access Control

**Advisory:** GHSA-v8x2-fjv7-8hjh

---

### Problems with this PR

**1. Overly Broad Check - Will Break Legitimate Edits**

The code compares the entire object:
```php
if ($originalData !== $newData && !$this->user->authorize('admin.super')) {
    throw new RuntimeException(...);
}
```

This blocks **ALL changes** by non-superusers, not just frontmatter changes. An editor making a legitimate content edit would be blocked.

**2. Check Happens Too Late**

The fix captures `$originalData` at line ~860, then the form data is applied to the object, then it checks permissions. The data is already on the object when the exception is thrown.

**3. Error Message Mismatch**

Says "edit frontmatter" but the code blocks any modification.

---

### Recommended Fix

The check should specifically target the `frontmatter` field in the incoming `$data`, placed **before** it gets applied:

```php
// Around line 867, BEFORE the existing frontmatter handling
if (!empty($data['frontmatter']) && !$this->user->authorize('admin.super')) {
    throw new RuntimeException(
        $this->admin::translate('PLUGIN_ADMIN.INSUFFICIENT_PERMISSIONS_FOR_TASK') . ' edit frontmatter.',
        403
    );
}

// Existing code continues...
if (!empty($data['frontmatter'])) {
    $frontmatter = (array) Yaml::parse($data['frontmatter']);
    unset($data['frontmatter']);
}
```

This:
- Blocks only frontmatter edits (not all edits)
- Checks before data is applied
- Is simpler and more targeted

---

**Do not merge as-is.** The fix would break legitimate editor functionality.

nakkouchtarek committed:
Fixed frontmatter permission patch

nakkouchtarek commented:
Took a look back at the initial patch, that was pretty messy, sorry for that 😅

I made two changes:

* I added your recommended fix on entering the function : 

```php
$callable = function (array $data, array $files, FlexObject $object) use ($form) {
    if (!empty($data['frontmatter']) && !$this->user->authorize('admin.super')) {
        throw new RuntimeException(
            $this->admin::translate('PLUGIN_ADMIN.INSUFFICIENT_PERMISSIONS_FOR_TASK') . ' edit frontmatter.',
            403
        );
    }
```

* I also removed the redundant permission check inside the support for expert mode part:

```php
// Support for expert mode.
if (str_ends_with($form->getId(), '-raw') && isset($data['frontmatter']) && is_callable([$object, 'frontmatter'])) {
    $object->frontmatter($data['frontmatter']);
    unset($data['frontmatter']);
}
```

Please do let me know if this is what you had in mind for the fix or if there's anything to change.

Cheers!
